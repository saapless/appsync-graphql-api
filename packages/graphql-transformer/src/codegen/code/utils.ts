/* eslint-disable security/detect-object-injection */

import { PropertyValue, tc } from "./ast";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}
