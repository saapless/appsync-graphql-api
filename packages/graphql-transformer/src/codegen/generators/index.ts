export { DynamoDbGenerator } from "./DynamoDBGenerator";
export { TypeGenerator } from "./type";
