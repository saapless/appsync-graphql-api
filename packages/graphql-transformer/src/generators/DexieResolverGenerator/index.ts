export { DexieResolverGenerator } from "./DexieResolversGenerator";
export { DexieResolverTypesGenerator } from "./DexieResolverTypesGenerator";
