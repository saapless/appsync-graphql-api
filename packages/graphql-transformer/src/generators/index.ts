export { ResolverGeneratorBase } from "./ResolverGeneratorBase";
export { DynamoDbGenerator } from "./DynamoDBGenerator";
export { NoneGenerator } from "./NoneGenerator";
export { SchemaTypesGenerator } from "./SchemaTypesGenerator";
export { ResolverTypesGenerator } from "./ResolverTypesGenerator";
