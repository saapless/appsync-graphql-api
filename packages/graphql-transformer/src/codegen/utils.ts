/* eslint-disable security/detect-object-injection */

import { KeyValue } from "../utils/types";
import { PropertyValue, tc } from "./code";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) return tc.chain("ctx", obj.ref);
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}
