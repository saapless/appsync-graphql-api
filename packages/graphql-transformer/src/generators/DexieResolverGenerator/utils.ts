import ts from "typescript";
import { addImport } from "../../utils";
import { KeyValue, ResolverDescriptor } from "../../context";

export function getWhereMethod(clause: string) {
  switch (clause) {
    case "eq":
    case "ref":
      return ts.factory.createIdentifier("equals");
    case "ne":
      return ts.factory.createIdentifier("notEqual");
    case "lt":
      return ts.factory.createIdentifier("below");
    case "le":
      return ts.factory.createIdentifier("belowOrEqual");
    case "gt":
      return ts.factory.createIdentifier("above");
    case "ge":
      return ts.factory.createIdentifier("aboveOrEqual");
    case "between":
      return ts.factory.createIdentifier("between");
    case "beginsWith":
      return ts.factory.createIdentifier("startsWith");
    default:
      throw new Error("Not implemented");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) {
    return ts.factory.createIdentifier(obj.ref);
  }

  if (obj.eq) {
    return typeof obj.eq === "string"
      ? ts.factory.createStringLiteral(obj.eq)
      : ts.factory.createNumericLiteral(obj.eq);
  }

  throw new Error("Invalid key value");
}

export function initQuery(descriptor: ResolverDescriptor) {
  return ts.factory.createVariableStatement(
    undefined,
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier("query"),
          undefined,
          undefined,
          ts.factory.createCallExpression(
            ts.factory.createPropertyAccessExpression(
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("ctx"),
                    ts.factory.createIdentifier("db")
                  ),
                  ts.factory.createIdentifier("where")
                ),
                undefined,
                [ts.factory.createStringLiteral(descriptor.operation.index ?? ":id")]
              ),
              ts.factory.createIdentifier("eq")
            ),
            undefined,
            [ts.factory.createStringLiteral("Task")]
          )
        ),
      ],
      ts.NodeFlags.Let
    )
  );
}

export function filterQuery() {
  return ts.factory.createIfStatement(
    ts.factory.createPropertyAccessExpression(
      ts.factory.createIdentifier("args"),
      ts.factory.createIdentifier("filter")
    ),
    ts.factory.createBlock(
      [
        ts.factory.createExpressionStatement(
          ts.factory.createBinaryExpression(
            ts.factory.createIdentifier("query"),
            ts.factory.createToken(ts.SyntaxKind.EqualsToken),
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("query"),
                ts.factory.createIdentifier("filter")
              ),
              undefined,
              [
                ts.factory.createCallExpression(
                  ts.factory.createIdentifier("filterExpression"),
                  undefined,
                  [
                    ts.factory.createPropertyAccessExpression(
                      ts.factory.createIdentifier("args"),
                      ts.factory.createIdentifier("filter")
                    ),
                  ]
                ),
              ]
            )
          )
        ),
      ],
      true
    ),
    undefined
  );
}

export function sortQuery() {
  return ts.factory.createIfStatement(
    ts.factory.createBinaryExpression(
      ts.factory.createPropertyAccessExpression(
        ts.factory.createIdentifier("args"),
        ts.factory.createIdentifier("sort")
      ),
      ts.factory.createToken(ts.SyntaxKind.EqualsEqualsEqualsToken),
      ts.factory.createStringLiteral("DESC")
    ),
    ts.factory.createBlock(
      [
        ts.factory.createExpressionStatement(
          ts.factory.createBinaryExpression(
            ts.factory.createIdentifier("query"),
            ts.factory.createToken(ts.SyntaxKind.EqualsToken),
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("query"),
                ts.factory.createIdentifier("reverse")
              ),
              undefined,
              []
            )
          )
        ),
      ],
      true
    ),
    undefined
  );
}

export function getQueryResult() {
  return ts.factory.createVariableStatement(
    undefined,
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier("result"),
          undefined,
          undefined,
          ts.factory.createAwaitExpression(
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("query"),
                ts.factory.createIdentifier("toArray")
              ),
              undefined,
              []
            )
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

export function initCreateItem(descriptor: ResolverDescriptor) {
  return [
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("id"),
            undefined,
            undefined,
            ts.factory.createBinaryExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createIdentifier("args"),
                  ts.factory.createIdentifier("input")
                ),
                ts.factory.createIdentifier("id")
              ),
              ts.factory.createToken(ts.SyntaxKind.QuestionQuestionToken),
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("ctx"),
                    ts.factory.createIdentifier("util")
                  ),
                  ts.factory.createIdentifier("uuid")
                ),
                undefined,
                []
              )
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("timestamp"),
            undefined,
            undefined,
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createNewExpression(ts.factory.createIdentifier("Date"), undefined, []),
                ts.factory.createIdentifier("toISOString")
              ),
              undefined,
              []
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("values"),
            undefined,
            undefined,
            ts.factory.createObjectLiteralExpression(
              [
                ts.factory.createSpreadAssignment(
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("args"),
                    ts.factory.createIdentifier("input")
                  )
                ),
                ts.factory.createPropertyAssignment(
                  ts.factory.createIdentifier("id"),
                  ts.factory.createIdentifier("id")
                ),
                ts.factory.createPropertyAssignment(
                  ts.factory.createIdentifier("createdAt"),
                  ts.factory.createBinaryExpression(
                    ts.factory.createPropertyAccessExpression(
                      ts.factory.createPropertyAccessExpression(
                        ts.factory.createIdentifier("args"),
                        ts.factory.createIdentifier("input")
                      ),
                      ts.factory.createIdentifier("createdAt")
                    ),
                    ts.factory.createToken(ts.SyntaxKind.QuestionQuestionToken),
                    ts.factory.createIdentifier("timestamp")
                  )
                ),
                ts.factory.createPropertyAssignment(
                  ts.factory.createIdentifier("updatedAt"),
                  ts.factory.createBinaryExpression(
                    ts.factory.createPropertyAccessExpression(
                      ts.factory.createPropertyAccessExpression(
                        ts.factory.createIdentifier("args"),
                        ts.factory.createIdentifier("input")
                      ),
                      ts.factory.createIdentifier("updatedAt")
                    ),
                    ts.factory.createToken(ts.SyntaxKind.QuestionQuestionToken),
                    ts.factory.createIdentifier("timestamp")
                  )
                ),
                ts.factory.createPropertyAssignment(
                  ts.factory.createIdentifier("__typename"),
                  ts.factory.createStringLiteral(descriptor.targetName)
                ),
              ],
              true
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createExpressionStatement(
      ts.factory.createAwaitExpression(
        ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(
            ts.factory.createPropertyAccessExpression(
              ts.factory.createIdentifier("ctx"),
              ts.factory.createIdentifier("db")
            ),
            ts.factory.createIdentifier("add")
          ),
          undefined,
          [ts.factory.createIdentifier("values")]
        )
      )
    ),
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("result"),
            undefined,
            undefined,
            ts.factory.createAwaitExpression(
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createIdentifier("ctx.db"),
                  ts.factory.createIdentifier("get")
                ),
                undefined,
                [ts.factory.createIdentifier("id")]
              )
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
  ];
}

export function initGetItem(descriptor: ResolverDescriptor) {
  return ts.factory.createVariableStatement(
    undefined,
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier("result"),
          undefined,
          undefined,
          ts.factory.createAwaitExpression(
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("ctx.db"),
                ts.factory.createIdentifier("get")
              ),
              undefined,
              [
                keyValue(
                  Array.isArray(descriptor.operation.key)
                    ? descriptor.operation.key[0]
                    : descriptor.operation.key
                ),
              ]
            )
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

export function initBulkGet() {
  return ts.factory.createVariableStatement(
    undefined,
    ts.factory.createVariableDeclarationList(
      [
        ts.factory.createVariableDeclaration(
          ts.factory.createIdentifier("result"),
          undefined,
          undefined,
          ts.factory.createAwaitExpression(
            ts.factory.createCallExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createIdentifier("db"),
                  ts.factory.createIdentifier("friends")
                ),
                ts.factory.createIdentifier("bulkGet")
              ),
              undefined,
              [
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createIdentifier("source"),
                  ts.factory.createIdentifier("keys")
                ),
              ]
            )
          )
        ),
      ],
      ts.NodeFlags.Const
    )
  );
}

export function initUpdateItem() {
  return [
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("id"),
            undefined,
            undefined,
            ts.factory.createPropertyAccessExpression(
              ts.factory.createPropertyAccessExpression(
                ts.factory.createIdentifier("args"),
                ts.factory.createIdentifier("input")
              ),
              ts.factory.createIdentifier("id")
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("updated"),
            undefined,
            undefined,
            ts.factory.createAwaitExpression(
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("ctx"),
                    ts.factory.createIdentifier("db")
                  ),
                  ts.factory.createIdentifier("update")
                ),
                undefined,
                [
                  ts.factory.createIdentifier("id"),
                  ts.factory.createObjectLiteralExpression(
                    [
                      ts.factory.createSpreadAssignment(
                        ts.factory.createPropertyAccessExpression(
                          ts.factory.createIdentifier("args"),
                          ts.factory.createIdentifier("input")
                        )
                      ),
                      ts.factory.createPropertyAssignment(
                        ts.factory.createIdentifier("updatedAt"),
                        ts.factory.createBinaryExpression(
                          ts.factory.createPropertyAccessExpression(
                            ts.factory.createPropertyAccessExpression(
                              ts.factory.createIdentifier("args"),
                              ts.factory.createIdentifier("input")
                            ),
                            ts.factory.createIdentifier("updatedAt")
                          ),
                          ts.factory.createToken(ts.SyntaxKind.QuestionQuestionToken),
                          ts.factory.createCallExpression(
                            ts.factory.createPropertyAccessExpression(
                              ts.factory.createNewExpression(
                                ts.factory.createIdentifier("Date"),
                                undefined,
                                []
                              ),
                              ts.factory.createIdentifier("toISOString")
                            ),
                            undefined,
                            []
                          )
                        )
                      ),
                    ],
                    true
                  ),
                ]
              )
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createIfStatement(
      ts.factory.createPrefixUnaryExpression(
        ts.SyntaxKind.ExclamationToken,
        ts.factory.createIdentifier("updated")
      ),
      ts.factory.createBlock(
        [
          ts.factory.createThrowStatement(
            ts.factory.createNewExpression(ts.factory.createIdentifier("Error"), undefined, [
              ts.factory.createStringLiteral("Record to update not found"),
            ])
          ),
        ],
        true
      ),
      undefined
    ),
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("result"),
            undefined,
            undefined,
            ts.factory.createAwaitExpression(
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createIdentifier("ctx.db"),
                  ts.factory.createIdentifier("get")
                ),
                undefined,
                [ts.factory.createIdentifier("id")]
              )
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
  ];
}

export function initDeleteItem() {
  return [
    ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList(
        [
          ts.factory.createVariableDeclaration(
            ts.factory.createIdentifier("record"),
            undefined,
            undefined,
            ts.factory.createAwaitExpression(
              ts.factory.createCallExpression(
                ts.factory.createPropertyAccessExpression(
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("ctx"),
                    ts.factory.createIdentifier("db")
                  ),
                  ts.factory.createIdentifier("get")
                ),
                undefined,
                [
                  ts.factory.createPropertyAccessExpression(
                    ts.factory.createIdentifier("args"),
                    ts.factory.createIdentifier("id")
                  ),
                ]
              )
            )
          ),
        ],
        ts.NodeFlags.Const
      )
    ),
    ts.factory.createIfStatement(
      ts.factory.createPrefixUnaryExpression(
        ts.SyntaxKind.ExclamationToken,
        ts.factory.createIdentifier("record")
      ),
      ts.factory.createBlock(
        [
          ts.factory.createThrowStatement(
            ts.factory.createNewExpression(ts.factory.createIdentifier("Error"), undefined, [
              ts.factory.createStringLiteral("Record does not exists"),
            ])
          ),
        ],
        true
      ),
      undefined
    ),
    ts.factory.createExpressionStatement(
      ts.factory.createAwaitExpression(
        ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(
            ts.factory.createPropertyAccessExpression(
              ts.factory.createIdentifier("ctx"),
              ts.factory.createIdentifier("db")
            ),
            ts.factory.createIdentifier("delete")
          ),
          undefined,
          [
            ts.factory.createPropertyAccessExpression(
              ts.factory.createIdentifier("record"),
              ts.factory.createIdentifier("id")
            ),
          ]
        )
      )
    ),
  ];
}

export function formatResult(descriptor: ResolverDescriptor, ast: ts.Node[], identifier?: string) {
  const props: ts.ObjectLiteralElementLike[] = [];

  if (descriptor.isEdge) {
    props.push(
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier("items"),
        ts.factory.createArrayLiteralExpression([], false)
      ),
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier("keys"),
        ts.factory.createIdentifier("result")
      )
    );
  } else {
    props.push(
      ts.factory.createPropertyAssignment(
        ts.factory.createIdentifier("items"),
        ts.factory.createIdentifier("result")
      )
    );
  }

  switch (descriptor.returnType) {
    case "connection": {
      addImport(
        ast,
        "@saapless/graphql-utils",
        ts.factory.createImportSpecifier(
          false,
          undefined,
          ts.factory.createIdentifier("formatConnection")
        )
      );
      return ts.factory.createCallExpression(
        ts.factory.createIdentifier("formatConnection"),
        undefined,
        [ts.factory.createObjectLiteralExpression(props)]
      );
    }
    case "result":
    default:
      return ts.factory.createIdentifier(identifier ?? "result");
  }
}
