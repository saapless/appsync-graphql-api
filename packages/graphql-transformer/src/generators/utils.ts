/* eslint-disable security/detect-object-injection */

import { FieldLoaderDescriptor, Key, KeyValue, LoaderDescriptor } from "../utils/types";
import { PropertyValue, tc } from "../codegen";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) {
    // `ctx.source` is typed as optional in Context

    if (obj.ref.startsWith("source")) {
      obj.ref = obj.ref.replace("source", "source?");
    }

    return tc.chain("ctx", obj.ref);
  }
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}

export function isFieldLoader(descriptor: LoaderDescriptor): descriptor is FieldLoaderDescriptor {
  return Object.hasOwn(descriptor, "fieldName") && Object.hasOwn(descriptor, "typeName");
}

export function parseKey(key: Key) {
  return tc.obj(
    Object.entries(key).reduce(
      (agg, [name, op]) => {
        if (op.ref || op.eq) {
          Object.assign(agg, { [name]: keyValue(op) });
        } else if (op.le) {
          Object.assign(agg, { [name]: tc.obj({ le: keyValue(op.le) }) });
        } else if (op.lt) {
          Object.assign(agg, { [name]: tc.obj({ lt: keyValue(op.lt) }) });
        } else if (op.ge) {
          Object.assign(agg, { [name]: tc.obj({ ge: keyValue(op.ge) }) });
        } else if (op.gt) {
          Object.assign(agg, { [name]: tc.obj({ gt: keyValue(op.gt) }) });
        } else if (op.between) {
          Object.assign(agg, {
            [name]: tc.obj({ between: tc.arr(...op.between.map((i) => keyValue(i))) }),
          });
        } else if (op.beginsWith) {
          Object.assign(agg, { [name]: tc.obj({ beginsWith: keyValue(op.beginsWith) }) });
        }
        return agg;
      },
      {} as Record<string, PropertyValue>
    )
  );
}
