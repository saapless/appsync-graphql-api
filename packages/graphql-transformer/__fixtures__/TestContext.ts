import { TransformerContext, TransformerContextConfig } from "../src/context";

interface TestingTransformerConfig extends TransformerContextConfig {
  allowPrint?: boolean;
}

export class TestContext extends TransformerContext {
  public allowPrint: boolean;
  public files: Map<string, string> = new Map();

  constructor(config: TestingTransformerConfig) {
    super(config);
    this.allowPrint = config.allowPrint ?? false;
    this.files = new Map();
  }

  public override createOutputDirectory(path: string) {
    return path;
  }

  public printFile(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printFile(filePath, content);
    }

    this.files.set(filePath, content);
  }

  public printScript(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printScript(filePath, content);
    }

    this.files.set(filePath, content);
  }
}
