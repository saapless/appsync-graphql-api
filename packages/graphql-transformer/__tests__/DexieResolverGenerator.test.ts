import { TestContext } from "../__fixtures__/TestContext";
import { DocumentNode } from "../src/definition";
import { DexieResolverGenerator } from "../src/generators/DexieResolverGenerator";
import { FieldLoaderDescriptor, printDefinitions } from "../src/utils";

const context = new TestContext({
  document: DocumentNode.fromSource(/* GraphQL */ `
    type Task {
      id: ID!
      title: String!
      done: Boolean!
      labels: LabelConnection
    }

    type Label {
      id: ID!
      name: String!
      taskId: string
      task: Task
    }

    type LabelConnection {
      edges: [LabelEdge!]!
      pageInfo: PageInfo
    }

    type LabelEdge {
      cursor: String
      node: Label
    }

    type TaskConnection {
      edges: [TaskEdge!]!
      pageInfo: PageInfo
    }

    type TaskEdge {
      cursor: String
      node: Task
    }

    type PageInfo {
      hasNextPage: Boolean!
      hasPreviousPage: Boolean!
      startCursor: String
      endCursor: String
    }

    input CreateTaskInput {
      id: ID
      title: String!
    }

    input UpdateTaskInput {
      id: ID!
      title: String
      done: Boolean
    }

    input CreateLabelInput {
      id: ID
      name: String!
    }

    input UpdateLabelInput {
      id: ID!
      name: String
    }

    type Query {
      getTask(id: ID!): Task
      listTasks: TaskConnection!
    }

    type Mutation {
      createTask(input: CreateTaskInput!): Task
      updateTask(input: UpdateTaskInput!): Task
      deleteTask(id: ID!): Task
      createLabel(input: CreateLabelInput!): Label
      updateLabel(input: UpdateLabelInput!): Label
      deleteLabel(id: ID!): Label
    }
  `),
  outputDirectory: "__test__",
});

const generator = new DexieResolverGenerator(context);

describe("DexieResolverGenerator", () => {
  it("generates getItem resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Query",
      fieldName: "getTask",
      action: {
        type: "getItem",
        key: { id: { ref: "args.id" } },
      },
      targetName: "Task",
      returnType: "result",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Query.getTask.ts")).toMatchSnapshot();
  });

  it("generates getItem resolver with nested key", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Query",
      fieldName: "getTask",
      action: {
        type: "getItem",
        key: { id: { ref: "args.id" }, __typename: { eq: "Task" } },
      },
      targetName: "Task",
      returnType: "result",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Query.getTask.ts")).toMatchSnapshot();
  });

  it("generates batchGetItems resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Label",
      fieldName: "edges",
      action: {
        type: "batchGetItems",
        key: { id: { ref: "source.keys" } },
      },
      targetName: "Label",
      returnType: "edges",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Query.listTasks.ts")).toMatchSnapshot();
  });

  it("generates queryItems resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Query",
      fieldName: "listTasks",
      action: {
        type: "queryItems",
        key: { id: { eq: "Task" } },
        index: "__typename",
      },
      targetName: "Task",
      returnType: "connection",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Query.listTasks.ts")).toMatchSnapshot();
  });

  it("generates createItem resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Mutation",
      fieldName: "createTask",
      action: {
        type: "putItem",
        key: { id: { ref: "args.input.id" } },
      },
      targetName: "Task",
      returnType: "result",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Mutation.createTask.ts")).toMatchSnapshot();
  });

  it("generates updateItem resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Mutation",
      fieldName: "updateTask",
      action: {
        type: "updateItem",
        key: { id: { ref: "args.input.id" } },
      },
      targetName: "Task",
      returnType: "result",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Mutation.updateTask.ts")).toMatchSnapshot();
  });

  it("generates deleteItem resolver", () => {
    const descriptor = {
      dataSource: "test",
      typeName: "Mutation",
      fieldName: "deleteTask",
      action: {
        type: "removeItem",
        key: { id: { ref: "args.id" } },
      },
      targetName: "Task",
      returnType: "result",
    } satisfies FieldLoaderDescriptor;

    const result = generator.generate(descriptor);
    expect(printDefinitions([result], "Mutation.deleteTask.ts")).toMatchSnapshot();
  });
});
