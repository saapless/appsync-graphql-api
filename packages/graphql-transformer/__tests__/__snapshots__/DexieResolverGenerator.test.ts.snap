// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DexieResolverGenerator generates batchGetItems resolver 1`] = `
"async (source, _, ctx) => {
    const result = await db.friends.bulkGet(source.keys);
    return result;
}
"
`;

exports[`DexieResolverGenerator generates createItem resolver 1`] = `
"async (_, args, ctx) => {
    const id = args.input.id ?? ctx.util.uuid();
    const timestamp = new Date().toISOString();
    const values = {
        ...args.input,
        id: id,
        createdAt: args.input.createdAt ?? timestamp,
        updatedAt: args.input.updatedAt ?? timestamp,
        __typename: "Task"
    };
    await ctx.db.add(values);
    const result = await ctx.db.get(id);
    return result;
}
"
`;

exports[`DexieResolverGenerator generates deleteItem resolver 1`] = `
"async (_, args, ctx) => {
    const record = await ctx.db.get(args.id);
    if (!record) {
        throw new Error("Record does not exists");
    }
    await ctx.db.delete(record.id);
    return record;
}
"
`;

exports[`DexieResolverGenerator generates getItem resolver 1`] = `
"async (_, args, ctx) => {
    const result = await ctx.db.get(id);
    return result;
}
"
`;

exports[`DexieResolverGenerator generates getItem resolver with nested key 1`] = `
"async (_, args, ctx) => {
    const result = await ctx.db.get(id);
    return result;
}
"
`;

exports[`DexieResolverGenerator generates queryItems resolver 1`] = `
"async (_, __, ctx) => {
    let query = ctx.db.where("__typename").eq("Task");
    if (args.filter) {
        query = query.filter(filterExpression(args.filter));
    }
    if (args.sort === "DESC") {
        query = query.reverse();
    }
    const result = await query.toArray();
    return formatConnection({ items: result });
}
"
`;

exports[`DexieResolverGenerator generates updateItem resolver 1`] = `
"async (_, args, ctx) => {
    const id = args.input.id;
    const updated = await ctx.db.update(id, {
        ...args.input,
        updatedAt: args.input.updatedAt ?? new Date().toISOString()
    });
    if (!updated) {
        throw new Error("Record to update not found");
    }
    const result = await ctx.db.get(id);
    return result;
}
"
`;
